### PR TITLE
feat: steady rate load generation for loadgen

### LINKED ISSUE #10444: apmsoak: add support for generating traffic bursts
`apmsoak` currently supports specifying a maximum event rate (eps), so we can verify that APM Server is able to handle a sustained workload. This also sets the burst size to `max(1000, 2*eps)`.

We should expand on this with support for configuring the burst size, to verify the server's ability to handle sustained bursty workloads. For example, it should be possible to configure `apmsoak` to send a burst of 10000 events every 5 seconds.

### PR DESCRIPTION
**Pull request by kyungeunni**
## Motivation/summary
Support for configuring the burst size and interval to verify the server's ability to handle sustained bursty workloads. 

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
- [x] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues
closes #10444

### CODE REVIEW

File: systemtest/loadgen/config/config.go [hunk 1]
```diff
@@ -52,29 +53,42 @@ func init() {
...
+				Config.Interval = t
 			case "m":
-				Config.MaxEPM = rateVal
+				Config.Interval = t * 60
 			default:
 				return fmt.Errorf(errStr, rate)
 			}
+			Config.EPS = burst / Config.Interval
 			return nil
 		})
```
Comment (axw) [inline systemtest/loadgen/config/config.go:91]: IMO this code has become sufficiently complicated that we should create a custom flag:

```go
type RateFlag struct {
        burst    int
        interval time.Duration
}

func (f *RateFlag) String() string {
        return fmt.Sprintf("%d/%s", f.burst, f.interval)
}

func (f *RateFlag) Set(s string) error {
        before, after, ok := strings.Cut(s, "/")
        if !ok || before == "" || after == "" {
                return fmt.Errorf("invalid rate %q, expected format burst/duration", s)
        }

        burst, err := strconv.Atoi(before)
        if err != nil {
                return fmt.Errorf("invalid burst %s in event rate: %w", before, err)
        }

        if !(after[0] >= '0' && after[0] <= '9') {
                after = "1" + after
        }
        interval, err := time.ParseDuration(after)
        if err != nil {
                return fmt.Errorf("invalid interval %q in event rate: %w", after, err)
        }
        if interval <= 0 {
                return fmt.Errorf("invalid interval %q, must be positive", after)
        }

        f.burst = burst
        f.interval = interval
        return nil
}
```

(May require some changes, I haven't fully tested that.)

**Review by axw — COMMENTED** [review]
Looks good! My comments are mostly about readability than anything. I think the functionality looks sound.

File: systemtest/loadgen/eventhandler/handler.go [hunk 2]
```diff
@@ -266,10 +266,45 @@ func (h *Handler) sendBatch(
...
+		err := h.sendHTTPRequest(ctx, b, baseTimestamp, randomBits)
+		if err != nil {
+			return 0, err
+		}
+		return len(b.events), nil
 	}
 
+	events := make([]event, len(b.events))
+	copy(events, b.events)
+	eventsLen := len(b.events)
```
[the remark this answers is not in the document]
Comment (kyungeunni) [reply systemtest/loadgen/eventhandler/handler.go:280]: `events` slice is reassigned later on ([line 295](https://github.com/elastic/apm-server/pull/10563/files#diff-a9132018cf11cb7b898a19b33d048de3f436b455881eca781ab0a98ae3d969ffR295)) 
Instead, I can just keep the last index and avoid copying all the events. 

Regarding `eventsLen`, yes we can do that. `len(b.events)` was used everywhere and I kind of preferred to have it as a named variable, but no strong opinion. Will update!

[the remark this answers is not in the document]
Comment (axw) [reply systemtest/loadgen/eventhandler/handler.go:280]: `events = events[n:]` doesn't modify the memory of the original `events` slice, it creates a new slice, offset by `n` elements from the beginning of the original slice. There's no need to copy just for that.

File: systemtest/loadgen/limiter.go [hunk 3]
```diff
@@ -18,24 +18,12 @@
...
 
 import (
-	"math"
-
 	"golang.org/x/time/rate"
 )
 
-func GetNewLimiter(epm int) *rate.Limiter {
-	if epm <= 0 {
+func GetNewLimiter(eps int, interval int) *rate.Limiter {
```
[the remark this answers is not in the document]
Comment (kyungeunni) [reply systemtest/loadgen/limiter.go:24]: Agreed to all 👍

File: systemtest/loadgen/eventhandler/handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kyungeunni) [reply systemtest/loadgen/eventhandler/handler.go:280]: Oh, I see. I must have misread about the underlying array can be affected when doing so. TIL, thanks!

File: systemtest/loadgen/config/config.go (hunk 1 above)
Comment (kyungeunni) [reply systemtest/loadgen/config/config.go]: It works like a charm 👍

File: systemtest/loadgen/limiter.go [hunk 4]
```diff
@@ -18,24 +18,21 @@
...
 import (
-	"math"
+	"time"
 
 	"golang.org/x/time/rate"
 )
 
-func GetNewLimiter(epm int) *rate.Limiter {
-	if epm <= 0 {
+// Return a rate limiter that is configured to rate limit burst per interval.
```
Comment (axw) [inline systemtest/loadgen/limiter.go:26]:
```suggestion
// GetNewLimiter returns a rate limiter that is configured to rate limit burst per interval.
```

It's conventional to start doc comments with the name of the method or type or whatever else you're documenting.

File: systemtest/loadgen/limiter.go [hunk 5]
```diff
@@ -18,24 +18,21 @@
...
-}
-
-func getBurstSize(eps int) int {
-	burst := eps * 2
-	// Allow for a batch to have 1000 events minimum
-	if burst < 1000 {
-		burst = 1000
-	}
-	return burst
+	eps := float64(burst) / float64(interval/time.Second)
```
Comment (axw) [inline systemtest/loadgen/limiter.go:36]:
```suggestion
	eps := float64(burst) / interval.Seconds()
```

File: testing/apmsoak/terraform.tfvars.example [hunk 6]
```diff
@@ -20,9 +20,9 @@ gcp_zone = "us-west2-b"
 # API Key for auth against the given server URL
 # apm_api_key = "<api_key>"
 
-# Max rate at which events will be sent to the APM server
-# Defaults to `4000eps` (i.e. 4000 events per second)
-# apm_loadgen_max_rate = "8000eps"
+# Event rate at which events will be sent to the APM server
+# Defaults to `4000/s` (i.e. 4000 events per second)
+# apm_loadgen_event_rate = "8000/s"
```
Comment (axw) [inline testing/apmsoak/terraform.tfvars.example:25]:
```suggestion
# apm_loadgen_event_rate = "4000/s"
```

The supposed default is out of sync with `variables.tf` and the comment above, let's fix it while we're here.

**Review by axw — APPROVED** [review]
LGTM! Just a few minor things

**Pull request merged**